src/Sylius/Bundle/CoreBundle/CatalogPromotion/CommandHandler/TerminateCatalogPromotionHandler.php

```diff
@@ -0,0 +1,54 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\CoreBundle\CatalogPromotion\CommandHandler;
+
+use InvalidArgumentException;
+use SM\Factory\FactoryInterface;
+use Sylius\Bundle\CoreBundle\CatalogPromotion\Command\RemoveCatalogPromotion;
+use Sylius\Bundle\CoreBundle\CatalogPromotion\Command\TerminateCatalogPromotion;
+use Sylius\Component\Core\Model\CatalogPromotion;
+use Sylius\Component\Promotion\Event\CatalogPromotionUpdated;
+use Sylius\Component\Promotion\Model\CatalogPromotionTransitions;
+use Sylius\Component\Promotion\Repository\CatalogPromotionRepositoryInterface;
+use Symfony\Component\Messenger\MessageBusInterface;
+use Symfony\Component\Messenger\Stamp\DispatchAfterCurrentBusStamp;
+
+final class TerminateCatalogPromotionHandler
+{
+    public function __construct(
+        private CatalogPromotionRepositoryInterface $catalogPromotionRepository,
+        private FactoryInterface $stateMachineFactory,
+        private MessageBusInterface $commandBus,
+        private MessageBusInterface $eventBus,
+    ) {
+    }
+
+    public function __invoke(TerminateCatalogPromotion $command): void
+    {
+        /** @var CatalogPromotion|null $catalogPromotion */
+        $catalogPromotion = $this->catalogPromotionRepository->findOneBy(['code' => $command->code]);
+
+        if (null === $catalogPromotion) {
+            throw new InvalidArgumentException('Catalog promotion with given code does not exist.');
+        }
+
+        $catalogPromotion->setEnabled(false);
+
+        $stateMachine = $this->stateMachineFactory->get($catalogPromotion, CatalogPromotionTransitions::GRAPH);
+        $stateMachine->apply(CatalogPromotionTransitions::TRANSITION_PROCESS);
+
+        $this->eventBus->dispatch(new CatalogPromotionUpdated($command->code), [new DispatchAfterCurrentBusStamp()]);
+        $this->commandBus->dispatch(new RemoveCatalogPromotion($command->code), [new DispatchAfterCurrentBusStamp()]);
```

Comment:
I see now 😄 It's quite shady for me, that we introduce a new term `termination` to reflect, that the **removal** of catalog promotions consists of two steps (disabling and an actual removal). Maybe just use the word `terminate` in Behat scenarios to make it consistent? Or do everything in one handler, reflecting **one** intention of an administrator - in the end, the administrator **wants to remove** the catalog promotion, the two-step process is just an implementation detail. 

Moreover, maybe we should make some saga/process manager here? It feels inconsistent, that we have 2 steps of the process, but the first command consists of two of them (disablement logic and dispatching the command of catalog promotion removal) - which feels like we have a command (intention) with 2 steps inside and another with 1 step instead of two commands with 1 step each :dancer: But it can be a subject of refactor later 🖖

Reply:
Refactoring done 🚀